hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/RecordStreamSigningUtils.java

```diff
@@ -0,0 +1,287 @@
+import com.hedera.services.stream.proto.SignatureFile;
+import com.hedera.services.stream.proto.SignatureObject;
+import com.swirlds.common.crypto.DigestType;
+import com.swirlds.common.crypto.HashingOutputStream;
+import com.swirlds.common.io.streams.SerializableDataOutputStream;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.security.InvalidKeyException;
+import java.security.KeyPair;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.SignatureException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.tuple.Pair;
+
+/**
+ * Utility class for signing record stream files
+ */
+public class RecordStreamSigningUtils {
+    /**
+     * Hidden constructor
+     */
+    private RecordStreamSigningUtils() {}
+
+    /**
+     * Supported stream version file
+     */
+    public static final int SUPPORTED_STREAM_FILE_VERSION = 6;
+
+    /**
+     * a messageDigest object for digesting entire stream file and generating entire record stream
+     * file hash
+     */
+    private static MessageDigest streamDigest;
+
+    /**
+     * a messageDigest object for digesting metaData in the stream file and generating metaData
+     * hash. Metadata contains: record stream version || HAPI proto version || startRunningHash ||
+     * endRunningHash || blockNumber, where || denotes concatenation
+     */
+    private static MessageDigest metadataStreamDigest;
+
+    /**
+     * Generates a signature file for the given record stream file
+     *
+     * @param signatureFileDestination the full path where the signature file will be generated
+     * @param streamFileToSign         the stream file to be signed
+     * @param keyPair                  the keyPair used for signing
+     * @param hapiVersion              the hapi protobuf version
+     * @return true if the signature file was generated successfully, false otherwise
+     */
+    public static boolean signRecordStreamFile(
+            @NonNull final Path signatureFileDestination,
+            @NonNull final Path streamFileToSign,
+            @NonNull final KeyPair keyPair,
+            @NonNull final String hapiVersion) {
+
+        Objects.requireNonNull(signatureFileDestination, "signatureFileDestination must not be null");
+        Objects.requireNonNull(streamFileToSign, "streamFileToSign must not be null");
+        Objects.requireNonNull(keyPair, "keyPair must not be null");
+        Objects.requireNonNull(hapiVersion, "hapiVersion must not be null");
+
+        try {
+            final var recordFile = streamFileToSign.toFile();
+            final int version = getRecordStreamVersion(recordFile);
```

Comment:
If you reorganize this then `getRecordStreamVersion` isn't needed.  Just read the record file.  It either works or it doesn't.  If it works, the version is returned _and you've got the`RecordStreamFile`_ you need.

Reply:
If I check the version first, I can exit faster if version is not even supported.

Reply:
Yes, but ... _command line tool_.  Will the user appreciate the millisecond or two you've saved for him?  Is that appreciation worth as much as, say, that which you'll get when someone who has to maintain this tool - possibly you yourself 6months from now - have to read it?  Just a suggestion anyway.  Like all my comments, I'm just trying here to set up a learning experience where we can learn from each other.  I'm good with whatever you want to do.

Reply:
This is also following the event stream sign way to check the supported version first. Really appreciate your comments by reviewing the code in details. I definitely learned from your reviews. Thanks!